src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1305 @@
+                var properties = new MetadataReferenceProperties(aliases: projectReference.Aliases, embedInteropTypes: projectReference.EmbedInteropTypes);
+                return await tracker.SkeletonReferenceCache.GetOrBuildReferenceAsync(
+                    tracker, this, properties, cancellationToken).ConfigureAwait(false);
+            }
+        }
+        catch (Exception e) when (FatalError.ReportAndPropagateUnlessCanceled(e, cancellationToken, ErrorSeverity.Critical))
+        {
+            throw ExceptionUtilities.Unreachable();
+        }
+    }
+
+    /// <summary>
+    /// Get a metadata reference for the project's compilation.  Returns <see langword="null"/> upon failure, which 
+    /// can happen when trying to build a skeleton reference that fails to build.
+    /// </summary>
+    public Task<MetadataReference?> GetMetadataReferenceAsync(
+        ProjectReference projectReference, ProjectState fromProject, CancellationToken cancellationToken)
+    {
+        try
+        {
+            // Get the compilation state for this project.  If it's not already created, then this
+            // will create it.  Then force that state to completion and get a metadata reference to it.
+            var tracker = this.GetCompilationTracker(projectReference.ProjectId);
+            return GetMetadataReferenceAsync(tracker, fromProject, projectReference, cancellationToken);
+        }
+        catch (Exception e) when (FatalError.ReportAndPropagateUnlessCanceled(e, cancellationToken, ErrorSeverity.Critical))
+        {
+            throw ExceptionUtilities.Unreachable();
+        }
+    }
+
+    /// <summary>
+    /// Undoes the operation of <see cref="WithFrozenSourceGeneratedDocument"/>; any frozen source generated document is allowed
+    /// to have it's real output again.
+    /// </summary>
+    public SolutionCompilationState WithoutFrozenSourceGeneratedDocuments(ProjectDependencyGraph dependencyGraph)
+    {
+        // If there's nothing frozen, there's nothing to do.
+        if (_frozenSourceGeneratedDocumentState == null)
+            return this;
+
+        var projectId = _frozenSourceGeneratedDocumentState.Id.ProjectId;
+
+        // Since we previously froze this document, we should have a CompilationTracker entry for it, and it should be a
+        // GeneratedFileReplacingCompilationTracker. To undo the operation, we'll just restore the original CompilationTracker.
+        var newTrackerMap = CreateCompilationTrackerMap(projectId, dependencyGraph);
+        Contract.ThrowIfFalse(newTrackerMap.TryGetValue(projectId, out var existingTracker));
+        var replacingItemTracker = existingTracker as GeneratedFileReplacingCompilationTracker;
+        Contract.ThrowIfNull(replacingItemTracker);
+        newTrackerMap = newTrackerMap.SetItem(projectId, replacingItemTracker.UnderlyingTracker);
+
+        return this.Branch(
+            // TODO(cyrusn): Is it ok to preserve the same solution here?
+            this.Solution,
```

Comment:
@jasonmalinowski  this was the case before this moved over here.  ebcause only the lower two values change, we're not actually changing the solution state.

Reply:
Yes, this is correct: this method is only changing what is being held for the source generators, so the underlying model doesn't change.